### PR TITLE
Throw when unsupported HTTPS options are set for HTTP/3

### LINKED ISSUE #34739: Throw when unsupported HTTPS options are set for HTTP/3

### PR DESCRIPTION
**Pull request by wtgodbe**
Fixes https://github.com/dotnet/aspnetcore/issues/34739. I'll open a backport to rc1 once approved.

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 1]
```diff
@@ -256,6 +256,10 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOpt
         /// <returns>The <see cref="ListenOptions"/>.</returns>
         public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOptionsSelectionCallback serverOptionsSelectionCallback, object state, TimeSpan handshakeTimeout)
         {
+            if (listenOptions.Protocols >= HttpProtocols.Http3)
```
Comment (wtgodbe) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:259]: This catches `Http1AndHttp2AndHttp3` also, should it only be checking for `Http3`?

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:259]: You can do this to check for HTTP/3: https://github.com/dotnet/aspnetcore/blob/204ec26bac3199176d70914b41721e69959eaf32/src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs#L208

I think we want to catch any usage of HTTP/3. Having it work for 1.1 & 2, but fail as soon as a HTTP/3 request comes in would be confusing.

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 2]
```diff
@@ -256,6 +256,10 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOpt
...
+            {
+                throw new NotSupportedException("UseHttps with ServerOptionsSelectionCallback is unsupported for HTTP/3");
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:261]:
```suggestion
                throw new NotSupportedException($"{nameof(UseHttps)} with {nameof(ServerOptionsSelectionCallback)} is not supported with HTTP/3.");
```

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 3]
```diff
@@ -283,6 +287,11 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, TlsHandsh
                 throw new ArgumentException($"{nameof(TlsHandshakeCallbackOptions.OnConnection)} must not be null.");
             }
 
+            if (listenOptions.Protocols >= HttpProtocols.Http3)
+            {
+                throw new NotSupportedException("UseHttps with TlsHandshakeCallbackOptions is unsupported for HTTP/3");
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:292]:
```suggestion
                throw new NotSupportedException($"{nameof(UseHttps)} with {nameof(TlsHandshakeCallbackOptions)} is not supported with HTTP/3.");
```

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:259]:
```suggestion
            if (listenOptions.Protocols.HasFlags(HttpProtocols.Http3))
```

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs [hunk 4]
```diff
@@ -510,6 +510,11 @@ private static bool IsWindowsVersionIncompatibleWithHttp2()
 
         internal static SslServerAuthenticationOptions CreateHttp3Options(HttpsConnectionAdapterOptions httpsOptions)
         {
+            if (httpsOptions.OnAuthenticate != null)
+            {
+                throw new NotSupportedException("The OnAuthenticate callback is unsupported for HTTP/3");
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:515]:
```suggestion
                throw new NotSupportedException($"The {nameof(HttpsConnectionAdapterOptions.OnAuthenticate)} callback is not supported with HTTP/3.");
```

**Review by Tratcher — APPROVED** [review]

**Review by JamesNK — COMMENTED** [review]
The string nameof changes probably won't compile by default. Also need to update test.

**Comment by JamesNK** [comment]
> I'll open a backport to rc1 once approved.

6.0 branch is fine. These changes aren't urgent.

**Review by JamesNK — APPROVED** [review]

**Review by adityamandaleeka — APPROVED** [review]

**Pull request merged**